pkg/volume/csi/csi_block.go

```diff
@@ -47,80 +51,61 @@ type csiBlockMapper struct {
 
 var _ volume.BlockVolumeMapper = &csiBlockMapper{}
 
-// GetGlobalMapPath returns a path (on the node) to a device file which will be symlinked to
+// GetGlobalMapPath returns a global map path (on the node) to a device file which will be symlinked to
 // Example: plugins/kubernetes.io/csi/volumeDevices/{volumeID}/dev
 func (m *csiBlockMapper) GetGlobalMapPath(spec *volume.Spec) (string, error) {
 	dir := getVolumeDevicePluginDir(spec.Name(), m.plugin.host)
 	klog.V(4).Infof(log("blockMapper.GetGlobalMapPath = %s", dir))
 	return dir, nil
 }
 
+// getStagingPath returns a staging path for a directory (on the node) that should be used on NodeStageVolume/NodeUnstageVolume
+// Example: plugins/kubernetes.io/csi/volumeDevices/staging/{volumeID}
+func (m *csiBlockMapper) getStagingPath() string {
+	sanitizedSpecVolID := kstrings.EscapeQualifiedNameForDisk(m.specName)
+	return path.Join(m.plugin.host.GetVolumeDevicePluginDir(csiPluginName), "staging", sanitizedSpecVolID)
+}
+
+// getPublishPath returns a publish path for a file (on the node) that should be used on NodePublishVolume/NodeUnpublishVolume
+// Example: plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}
```

Comment:
If multiple containers use this, it will collide.

Reply:
@saad-ali 

Thank you for your comment.

Is my understanding right that you are suggesting that we should use ```spec.PersistentVolume.Name``` instead of ```kstrings.EscapeQualifiedNameForDisk(m.specName)```, as it is done in ```makeDeviceMountPath``` for filesystem volume?

Then, it will be like as below:
- block: ```/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/{pvname}```
             ```/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}```
- filesystem: ```/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount```


```m.specName``` is originally used there, because in-tree block volume code use it for ```GlobalMapPath``` and ```PodDeviceMapPath```.
- [GlobalMapPath](https://github.com/kubernetes/kubernetes/pull/68635/files#diff-ed824558531f8d1f616bc2f976266b64R54) : ```/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/{volumeID}/dev```
- [PodDeviceMapPath](https://github.com/kubernetes/kubernetes/pull/68635/files#diff-ed824558531f8d1f616bc2f976266b64R77): ```/var/lib/kubelet/pods/{podUid}/volumeDevices/kubernetes.io~csi/{volumeID}```

However, as explained in https://github.com/kubernetes/kubernetes/pull/68635#issuecomment-423259818, for block volume, we choose not to let csi driver handle ```GlobalMapPath``` and ```PodDeviceMapPath``` directly, instead csi driver make block volume visible to another path (```target_path```), then k8s (```ioutil.MapBlockVolume```) make ```target_path visible``` from ```GlobalMapPath``` and ```PodDeviceMapPath```. So, there is no restriction for using ```m.specName``` for ```target_path``` and ```staging_target_path```.

@bswartz @vladimirvivien 
Any concern?

Reply:
@saad-ali 

I checked the log and code again and found out that block codes already use the same ID to filesystem codes. I mean ```{pvname}``` should be the same to ```{volumeID}```, like ```pvc-a618a282-e929-11e8-bb66-525400b854f0```, so it should not collide.

Is it ok to just update the comment to make them look same?

Reply:
Would `spec.PersistentVolume.Name` collide if the PVC is used by multiple pods? Ideally the `target_path` should be unique per pod (have the pod ID in it) right? Otherwise if 2 pods use the same volume on the same node, each pod will have it's volume setup code running in parallel. 

I'm basing this on my understanding of mount, so it may well be different for block. If so I'd like to understand why?